runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -1158,6 +1158,29 @@ local loclist_type_map = {
   [DiagnosticSeverity.Hint] = 'I',
 }
 
+
+--- Clear diagnotics and diagnostic cache
+---
+--- Handles saving diagnostics from multiple clients in the same buffer.
+---@param client_id number
+---@param buffer_client_map table map of buffers to active clients
```

Comment:
Is this supposed to modify the client map in place? I'd be surprised about that if I were the consumer.

Reply:
Not if we want it to be a more general function, I changed it such that we call this:

```lua
    lsp.diagnostic.reset(client_id, all_buffer_active_clients)
    all_client_active_buffers[client_id] = nil
    for _, client_ids in pairs(all_buffer_active_clients) do
      client_ids[client_id] = nil
    end
```

In both `dispatch.on_exit` and `client.stop`. on_detach also removes the buffer from `all_buffer_active_clients`, but I think it's also a good idea in these two places?